src/coreclr/tools/aot/ILCompiler.Diagnostics/PerfMapWriter.cs

```diff
@@ -72,6 +55,96 @@ public static void Write(string perfMapFileName, int perfMapFormatVersion, IEnum
+        }
+
+        public static byte[] PerfMapV1SignatureHelper(IEnumerable<AssemblyInfo> inputAssemblies, TargetDetails details)
+        {
+            IEnumerable<AssemblyInfo> orderedInputs = inputAssemblies.OrderBy(asm => asm.Name, StringComparer.OrdinalIgnoreCase);
+            List<byte> inputHash = new List<byte>();
+            foreach (AssemblyInfo inputAssembly in orderedInputs)
+            {
+                inputHash.AddRange(inputAssembly.Mvid.ToByteArray());
+            }
+
+            PerfmapTokensForTarget targetTokens = TranslateTargetDetailsToPerfmapConstants(details);
+
+            byte[] buffer = new byte[12];
+            if (!BitConverter.TryWriteBytes(buffer.AsSpan(0, sizeof(uint)), (uint)targetTokens.OperatingSystem)
+                || !BitConverter.TryWriteBytes(buffer.AsSpan(4, sizeof(uint)), (uint)targetTokens.Architecture)
+                || !BitConverter.TryWriteBytes(buffer.AsSpan(8, sizeof(uint)), (uint)targetTokens.Abi))
+            {
+                throw new InvalidOperationException();
+            }
+
+            if (!BitConverter.IsLittleEndian)
+            {
+                buffer.AsSpan(0, sizeof(uint)).Reverse();
+                buffer.AsSpan(4, sizeof(uint)).Reverse();
+                buffer.AsSpan(8, sizeof(uint)).Reverse();
+            }
+
+            inputHash.AddRange(buffer);
+            byte[] hash = MD5.HashData(inputHash.ToArray());
+
+            return hash;
+        }
+
+        internal record struct PerfmapTokensForTarget(PerfMapOSToken OperatingSystem, PerfMapArchitectureToken Architecture, PerfMapAbiToken Abi);
+
+        private static PerfmapTokensForTarget TranslateTargetDetailsToPerfmapConstants(TargetDetails details)
+        {
+            PerfMapOSToken osToken = details.OperatingSystem switch
+            {
+                TargetOS.Unknown => PerfMapOSToken.Unknown,
+                TargetOS.Windows => PerfMapOSToken.Windows,
+                TargetOS.Linux => PerfMapOSToken.Linux,
+                TargetOS.OSX => PerfMapOSToken.OSX,
+                TargetOS.FreeBSD => PerfMapOSToken.FreeBSD,
+                TargetOS.NetBSD => PerfMapOSToken.NetBSD,
+                TargetOS.SunOS => PerfMapOSToken.SunOS,
+                TargetOS.WebAssembly => PerfMapOSToken.WebAssembly,
+                _ => throw new NotImplementedException(details.OperatingSystem.ToString())
+            };
+
+            PerfMapAbiToken abiToken = details.Abi switch
+            {
+                TargetAbi.Unknown => PerfMapAbiToken.Unknown,
+                TargetAbi.CoreRT => PerfMapAbiToken.CoreRT,
+                TargetAbi.CoreRTArmel => PerfMapAbiToken.CoreRTArmel,
+                TargetAbi.Jit => PerfMapAbiToken.Jit,
+                TargetAbi.CppCodegen => PerfMapAbiToken.CppCodegen,
+                _ => throw new NotImplementedException(details.Abi.ToString())
```

Comment:
I would take it as an opportunity to get rid of nonsensical names and never possible values.

1:1 mapping will be super confusing once we decide to clean up unused things and rename things that don't have a good name in the type system. Perfmap format is a stable format so it will be stuck with these decisions forever. They cannot be easily cleaned up.

```suggestion
            PerfMapAbiToken abiToken = details.Abi switch
            {
                TargetAbi.Unknown => PerfMapAbiToken.Unknown,
                TargetAbi.CoreRT => PerfMapAbiToken.Default,
                TargetAbi.CoreRTArmel => PerfMapAbiToken.Armel,
                _ => throw new NotImplementedException(details.Abi.ToString())
```